contracts/rust/src/light_client.rs

```diff
@@ -0,0 +1,405 @@
+                h: srs.h,
+                beta_h: srs.beta_h,
+                powers_of_h: vec![srs.h, srs.beta_h],
+            }
+        };
+        let (pk, _) = hotshot_state_prover::preprocess::<STAKE_TABLE_CAPACITY>(&srs)
+            .expect("Fail to preprocess state prover circuit");
+        let (proof, pi) =
+            hotshot_state_prover::generate_state_update_proof::<_, _, _, _, STAKE_TABLE_CAPACITY>(
+                &mut self.rng,
+                &pk,
+                &self.st,
+                &bit_vec,
+                &sigs,
+                &self.state,
+                &self.threshold,
+            )
+            .expect("Fail to generate state proof");
+        (pi, proof)
+    }
+
+    /// Returns the `LightClientState` for solidity
+    pub fn get_state(&self) -> ParsedLightClientState {
+        // The ugly conversion due to slight difference of `LightClientState` in solidity containing `threshold`
+        let pi = vec![
+            u256_to_field(self.threshold),
+            F::from(self.state.view_number as u64),
+            F::from(self.state.block_height as u64),
+            self.state.block_comm_root,
+            self.state.fee_ledger_comm,
+            self.state.stake_table_comm.0,
+            self.state.stake_table_comm.1,
+            self.state.stake_table_comm.2,
+        ];
+        let pi: PublicInput<F> = pi.into();
+        pi.into()
+    }
+
+    /// Returns the (bytes32 votingStakeTableComm, bytes32 frozenStakeTableComm) used in contract
+    pub fn get_stake_table_comms(&self) -> (H256, H256) {
+        let (bls_key_comm, schnorr_key_comm, amount_comm) =
+            self.st.commitment(SnapshotVersion::EpochStart).unwrap();
+        let frozen_st_comm = utils::keccak256(
+            abi::encode_packed(&[
+                Token::Uint(field_to_u256(bls_key_comm)),
+                Token::Uint(field_to_u256(schnorr_key_comm)),
+                Token::Uint(field_to_u256(amount_comm)),
+            ])
+            .unwrap(),
+        );
+
+        let (bls_key_comm, schnorr_key_comm, amount_comm) =
+            self.st.commitment(SnapshotVersion::LastEpochStart).unwrap();
+        let voting_st_comm = utils::keccak256(
+            abi::encode_packed(&[
+                Token::Uint(field_to_u256(bls_key_comm)),
+                Token::Uint(field_to_u256(schnorr_key_comm)),
+                Token::Uint(field_to_u256(amount_comm)),
+            ])
+            .unwrap(),
+        );
+
+        (voting_st_comm.into(), frozen_st_comm.into())
+    }
+
+    // return a dummy commitment value
+    fn new_dummy_comm(&mut self) -> F {
+        F::rand(&mut self.rng)
+    }
+}
+
+/// Helper function for test
+pub(crate) fn key_pairs_for_testing<R: CryptoRng + RngCore>(
+    num_validators: usize,
+    prng: &mut R,
+) -> (Vec<BLSVerKey>, Vec<(SchnorrSignKey, SchnorrVerKey)>) {
+    let bls_keys = (0..num_validators)
+        .map(|_| {
+            BLSOverBN254CurveSignatureScheme::key_gen(&(), prng)
```

Comment:
Why do we use the BN curve if HotShot replicas signature will be computed over the BLS12381 curve?

Reply:
I copied this from @mrain 's testing code in `HotShot/crates/hotshot-state-prover/lib.rs`, and he used that because currently consensus still use BN curve (which I believe is a legacy reason, I think we chose for forward compatibility reason in the future we might directly creating proofs over these BLS signatures)

Reply:
I see, thanks for the clarification. I did not know consensus currently uses the BN curve.